docs/design/Ratify v2 Architecture Proposal.md

```diff
@@ -0,0 +1,262 @@
+      - [Deisgn considerations](#deisgn-considerations)
+      - [Plugin Framework Example](#plugin-framework-example)
+    - [Poposed Repository Layout](#poposed-repository-layout)
+    - [Proposed Milestones](#proposed-milestones)
+    - [Q\&A](#qa)
+
+## Background
+Ratify has reached its first major release in over a year and has been widely adopted by users at scale. However, there are some know limitations in the current design and implementations of Ratify v1 that make it difficult to support more user scenarios/new features and enhance performance. These constraints make it challenging to attract more users and contributors. This is a proposal to address these limitations and improve the overall architecture of Ratify in the next major release, Ratify v2.
+
+## Previous Discussions
+In a previous [discussion](https://github.com/ratify-project/ratify/pull/1905), we have identified some areas that need to be improved in Ratify v2, which includes but not limited to: system refactoring, deprecations, document update and unit test improvement. This proposal will focus on the system refactoring part, specifically the overall architecture of Ratify v2 including new plugin framework and decoupling of core components.
+
+## Goals
+The goal of this proposal aligns with the goal in the previous discussion, specifically:
+1. Refactor the code to make it more modular and maintainable.
+2. Improve the performance of Ratify by optimizing the core components and making it easier to scale.
+3. Reduce the dependencies of Ratify so that it is easier to maintain and upgrade.
+4. Make it easier for contributors to develop and test new features.
+
+## Current Architecture of Ratify v1
+![img](../img/architecture/ratify-v1.png)
+
+### Limitations
+1. Ratify only supports CLI and K8s scenarios. But actually current Ratify is only well implemented/designed for K8s scenarios. There are some feature gaps between CLI and K8s implementations. To support more scenarios, like containerd plugins or docker plugins, or just a library for downstream services, we need to make Ratify to be more extensible.
+2. Ratify was designed to validate security metadata of any given artifacts. But right now Ratify is mainly focusing on the K8s scenarios with images being verified.
+3. The plugin framework was designed to separate built-in and external plugins. Built-in plugins run in the same process of Ratify while each external plugin would be executed in a new subprocess. Therefore, external plugins cannot share the in-momory cache with the main process, which may result to performance degradation, data inconsistency, race condition and security vulnerabilties.
+4. The built-in plugins and authentication for different cloud providers are part of the main Ratify repository, which introduces a significant number of dependencies. This issue will become even more pronounced as additional cloud provider and new plugin implementation are added in the future.
```

Comment:
Hi @binbin-li , after the refactoring, will auth providers be in Ratify core, have we called this out already?

Reply:
Ratify core will just define the Store interface without its implementation. The auth provider will belong to oras implementation repo.

Reply:
Actually my plan is to have ratify-go define `ReferrerStore` and `AuthProvider` interfaces. The authProvider implementation will reside in the entrypoint repos because KMP/Verifier will also need similar authProvider.